### ISSUE TITLE
ThreadedResolver--and twisted.internet.thread--assumes that the reactor is the installed reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3591]
|<img alt="gary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gary reported|
|-|-|
|Trac ID|trac#3591|
|Type|defect|
|Created|2008-12-23 21:13:15Z|
|Branch|https://github.com/twisted/twisted/tree/threaded-resolver-reactor-3591|

twisted.internet.base.ThreadedResolver appears to go out of its way to support working with the reactor on which it was installed--particularly, one that is not necessarily the reactor installed in twisted.internet.reactor.  However, in getHostByName, it uses twisted.internet.thread.deferToThread which does not honor self.reactor.

I worked around this by making a deferToThreadWithReactor function, and subclassing ThreadedResolver to use it, and putting an instance of my resolver subclass on my reactor instance.  It would be nice if that were unnecessary, because the resolver worked with its own reactor.  Perhaps twisted.internet.thread could have alternate functions that accept a reactor?  Then the current function names would defer to them, simply passing in the installed reactor explicitly?

I realize this is an odd edge case. But the Twisted code seems to imply that what I expected/wanted was the desired behavior: otherwise why instantiate ThreadedResolver with the reactor?

FWIW, I'm operating within another program (bzr) in variable circumstances (might be within a one-off RCS command, or might be within a long-running bzr server process).

<details><summary>Searchable metadata</summary>

```
trac-id__3591 3591
type__defect defect
reporter__gary gary
priority__normal normal
milestone__ 
branch__branches_threaded_resolver_reactor_3591 branches/threaded-resolver-reactor-3591
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1230066795000000 1230066795000000
changetime__1232632919000000 1232632919000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Thanks for reporting this.  I want to assure you that you have interpreted the intent correctly, and that things like `ThreadedResolver` which have a reactor attribute should use it exclusively, and never touch the "installed" reactor.  You've identified an important bug.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

`twisted.internet.threads.deferToThreadPool` exists now.  You can't just trivially switch `ThreadedResolver` over to it, though, because it requires a threadpool object.  The current implementation is able to use the reactor's threadpool by accessing an undocumented (notionally private) attribute of the reactor (`threadpool`).  I assume that your `deferToThreadWithReactor` does something similar.  This isn't a good solution.  One possible solution which may be better is to make the reactor's threading API amenable to this sort of usage.

**Comment by twisted-trac** [comment]
|<img alt="gary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|gary commented|
|-|-|

No, I don't need to access the thread pool directly.  Here's the pertinent snippet of what I have done locally:

# replaces twisted.internet.thread equivalent
def _putResultInDeferred(reactor, deferred, f, args, kwargs):
    """
    Run a function and give results to a Deferred.
    """
    try:
        result = f(*args, **kwargs)
    except:
        f = failure.Failure()
        reactor.callFromThread(deferred.errback, f)
    else:
        reactor.callFromThread(deferred.callback, result)

# would be a proposed addition.  deferToThread could use it
def deferToThreadInReactor(reactor, f, *args, **kwargs):
    """
    Run function in thread and return result as Deferred.
    """
    d = defer.Deferred()
    reactor.callInThread(_putResultInDeferred, reactor, d, f, args, kwargs)
    return d

# uses its own reactor for the threaded calls, unlike Twisted's
class ThreadedResolver(twisted.internet.base.ThreadedResolver):
    def getHostByName(self, name, timeout = (1, 3, 11, 45)):
        if timeout:
            timeoutDelay = reduce(operator.add, timeout)
        else:
            timeoutDelay = 60
        userDeferred = defer.Deferred()
        lookupDeferred = deferToThreadInReactor(
            self.reactor, socket.gethostbyname, name)
        cancelCall = self.reactor.callLater(
            timeoutDelay, self._cleanup, name, lookupDeferred)
        self._runningQueries[lookupDeferred] = (userDeferred, cancelCall)
        lookupDeferred.addBoth(self._checkTimeout, name, lookupDeferred)
        return userDeferred

...Then I instantiate the new ThreadedResolver and slam it on my reactor after I create it.  

You'll notice that the changes are very small.  Using this approach, I would expect ``deferToThread`` to then look something like this:

def deferToThread(f, *args, **kwargs):
    """
    Run function in thread and return result as Deferred.
    """
    return deferToThreadInReactor(twisted.internet.reactor, f, *args, **kwargs)

Maybe I'm missing where I would need access to the thread pool.  ATM my fix does not require access to private APIs AFAIK.

(Sorry for slow reply: holidays)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Just pointing out that `deferToThread` has changed a bit since the version you're looking at:

 http://twistedmatrix.com/trac/browser/trunk/twisted/internet/threads.py?rev=24507#L54

**Comment by twisted-trac** [comment]
|<img alt="gary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|gary commented|
|-|-|

Ah, heh, that would be what I'm missing.

While I agree that the fact that the current implementation is using a private attribute is nasty, it is related only by spatial proximity to the bug I'm reporting.  My idea was simply to add additional parallel functions in the threads module that allow passing in a reactor explicitly; and then change ThreadedResolver.getHostByName to use the new explicit-reactor functions.  That change doesn't touch the code to which you refer--or at most just shuffles it around.

Perhaps you have a larger change in mind that would address both problems simultaneously?

I also notice that my code has been wiki-mangled somehow.  Sorry, should have checked what the parser was.  If you want me to re-enter it more legibly let me know.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

I get the gist of your code, so don't worry about re-entering it.

Perhaps `ThreadedResolver` can just be switched to `deferToThreadPool`?  That API already allows a reactor to be passed in.  The only part of that solution which isn't obvious to me is whether `ThreadedResolver` should then pass in the reactor's threadpool or if it should make its own threadpool to pass in.

**Comment by twisted-trac** [comment]
|<img alt="gary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|gary commented|
|-|-|

Replying to [exarkun](#note_7):
> I get the gist of your code, so don't worry about re-entering it.

Cool.

> Perhaps `ThreadedResolver` can just be switched to `deferToThreadPool`?  That API already allows a reactor to be passed in.  

+1

> The only part of that solution which isn't obvious to me is whether `ThreadedResolver` should then pass in the reactor's threadpool or if it should make its own threadpool to pass in.

Eh, I can see the dilemma, I think, but I lean towards using the reactor's.  Fewer knobs to configure.  If it turns out there's a real need for a separate threadpool, change it when the need is discovered.  $.02

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [gary](#note_8):

> Eh, I can see the dilemma, I think, but I lean towards using the reactor's.  Fewer knobs to configure.  If it turns out there's a real need for a separate threadpool, change it when the need is discovered.  $.02

Strong agreement from me.  Let's just pass the reactor's threadpool.

And actually, if there's ever a need for a separate threadpool, I think that rather than adding it we should address the design wart that is the divergence between `ThreadPool` and `IReactorThreads` rather than just add an additional argument.  Not that we don't occasionally need distinct threadpools for different subsystems, it's just that they ought to be different implementations of _some_ common API, which a reactor can also provide.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25952]) Branching to 'threaded-resolver-reactor-3591'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

Okay

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_12"></a>|@mwhudson set owner to @exarkun|
|-|-|

This is a nice branch, good work.  One question:

suggestThreadPoolSize has changed in behaviour; in trunk if there is no threadpool when it is called, it has no effect, whereas in the branch it causes the threadpool to be created and then adjusted.

On writing this down, it sounds like a good idea -- having suggestThreadPoolSize do nothing at all sometimes doesn't seem sane!  But maybe it should be tested.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

Indeed, testing that behavior is a good idea.  Did that in r26103.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_14"></a>|@mwhudson set owner to @exarkun|
|-|-|

Cool beans.  Land it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [26105]) Merge threaded-resolver-reactor-3591

Author: exarkun, itamarst
Reviewer: mwhudson
Fixes: [#3591](https://github.com/twisted/twisted/issues/3591)

Add `IReactorThreads.getThreadPool` and convert `ThreadedResolver` to use it and
the `deferToThreadPool` helper so that when doing DNS lookups, only the reactor
passed to `ThreadedResolver` is used, not the global reactor (in case they happen
to be different).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
